multiraft/multiraft.go

```diff
@@ -871,3 +912,44 @@ func (s *state) handleWriteResponse(response *writeResponse, readyGroups map[uin
 		}
 	}
 }
+
+// processLease unmarshals the given command and, in case a leader lease
+// request is contained within, updates the group with the expiration date
+// constructed from the current wall time and the lease request's duration.
+// This causes local representant of the group to have its outgoing votes
+// dropped for the duration specified in the request.
+//
+// The mechanism for leader leases is unfortunately somewhat involved.
+// Leases are initiated on the range level by proposing a command to Raft
+// that contains the prospective lease holder's local expiration timestamp
+// and a duration (which is used by recipients of that message to construct
+// their local timestamps).
+// For this to work reliably, it is vital that the lease expires first at
+// the lease holder, and only begins when the lease command commits.
+// While a lease is active, the node may not elect a new leader, and this
+// is implemented in MultiRaft by dropping voting-related messages.
+// However, the amount of abstraction between Raft, MultiRaft and the Range
+// makes it cumbersome to find out where a command originated from.
+// For that reason, here we simply use the local clock plus the expiration
+// from the lease message before letting the client know the lease has been
+// committed to figure out the duration for which we won't participate in
+// elections. Since the grantee will only believe it has the lease once it
+// learns that the command committed, and since the range is taking its local
+// walltime strictly before we even see the command here, that only adds an
+// unecessary extra period of time in which nodes don't vote but does not
+// impact correctness.
+//
+// TODO(tschottdorf) having to unmarshal every single command just for leader
+// leases is horrible. Should do this in a way that's closer to MultiRaft.
```

Comment:
In addition to the performance implications, it's also unfortunate that this introduces a dependency from multiraft to the cockroach-specific protos. There's no easy way around it though so we can live with it until we rework this to avoid unmarshalling.